[#26] 요청에 대한 알림 추가 기능 구현

msugo1 opened the pull request:
## 📖 Enum 관련

### 📑  `AlarmTypes` 클래스 추가

- `Alarm 관련 타입 상수`를 담을 클래스

- DB에 저장 시 정수로, DB에서 꺼낼 때는 정식 명칭을 가진 상수로의 `자동변환`을 위해 `TypeHandler` 구현 및 MyBatisConfig에 추가

## 📖 Mapper 계층 관련

### 📑 `AlarmMapper` 인터페이스 추가

- 알람 추가 기능에 대한 로직 처리 후 이를 DB에 반영하기 위한 `insertAlarm` 메소드 추가

### 📑 `AlarmMapper.xml` 추가

- 위의 사항을 처리할 쿼리문 작성

## 📖 Service 계층 관련

### 📑 `AlarmService` 클래스 추가

- 알람 추가 기능 요청에 대한 비즈니스 로직을 처리할 `addAlarm` 메소드 추가

- 알람 추가 요청을 받으면 보내는 대상과 받을 대상 그리고 알람 타입을 매핑해서 `Alarm 클래스의 인스턴스 생성` 후 Mapper로 넘김

### 📑 FriendService

- `AlarmService 주입` 및 알람 추가가 필요한 메소드`(친구추가 요청, 친구수락 요청)`들에 `addAlarm코드 추가`

## 📖 테스트 관련

- 전체적인 내용을 포스트맨을 활용해 테스트

Commit:
[#25] 사용자 차단 목록 불러오기 기능 구현

Model 관련

* BlockUserListRequest 클래스 추가

- 현재 사용자 아이디, 차단 목록 페이지 위치, 그리고 한 페이지에
  출력할 아이디 개수의 제한 숫자를 멤버 변수로 갖는 인스턴스를 생성할
  클래스 추가

Controller 계층 관련

* BlockUserController

- 현재 로그인 중인 사용자의 차단 사용자 목록을 불러오는 요청을 받을
  `getBlockUserList` 메소드 추가

Service 계층 관련

* BlockUserService

- 현재 사용자의 차단 목록을 불러오는 비즈니스 로직을 처리할
  `getBlockUserList` 메소드 추가

- `@Transactional(readOnly=true)를 붙여서 Slave DB에서 내용을 읽어오도록
  설정

Mapper 계층 관련

* BlockUserMapper

- DB에서 차단 사용자 목록을 불러 올 `getBlockUserList` 메소드 추가

* BlockUserMapper.xml

- 위의 내용을 처리할 쿼리문 추가

Util 관련

* Pagination

- 차단 목록에 대한 max per page 멤버변수 추가

테스트 관련

- 포스트맨을 사용해 위의 로직 테스트

Commit:
[#26] 요청에 대한 알림 추가 기능 구현

Enum 관련

* AlarmTypes 클래스 추가

- Alarm 관련 타입 상수를 담을 클래스

- DB에 저장 시 정수로, DB에서 꺼낼 때는 정식 명칭을 가진 상수로의 자동
  변환을 위해 `TypeHandler` 구현 및 MyBatisConfig에 추가

Mapper 계층 관련

* AlarmMapper 인터페이스 추가

- 알람 추가 기능에 대한 로직 처리 후 이를 DB에 반영하기 위한
  `insertAlarm` 메소드 추가

* AlarmMapper.xml 추가

- 위의 사항을 처리할 쿼리문 작성

Service 계층 관련

* AlarmService 클래스 추가

- 알람 추가 기능 요청에 대한 비즈니스 로직을 처리할 `addAlarm` 메소드 추가

- 알람 추가 요청을 받으면 보내는 대상과 받을 대상 그리고 알람 타입을
  매핑해서 `Alarm 클래스의 인스턴스 생성` 후 Mapper로 넘김

* FriendService

- AlarmService 주입 및 알람 추가가 필요한 메소드`(친구추가 요청, 친구수락 요청)들에addAlarm 코드 추가

테스트 관련

- 전체적인 내용을 포스트맨을 활용해 테스트

msugo1 commented:
### 질문

#### 1. `@Transactional` 필요 여부?
- 친구 추가 요청을 보낼 때 요청을 먼저 DB에 저장한 후 알람이 DB에 저장이 되는 구조인데요. 서로 한 트랜잭션에서 처리되어야 하니 `PROPAGATION` 부분은 건드릴 필요가 없을 듯 한데요. 문득 든 생각이 **어차피 친구 추가 요청을 보내는데 실패하면 예외가 발생할텐데 그 다음 부분인 알람 저장하기 부분을 위해서 트랜잭션을 붙일 필요가 있을까?** 였습니다. 일단 붙여주지 않았는데요. 혹시 해당 로직에 `@Transactional`을 붙여주지 않으면 멀티 스레딩 환경에서 문제가 생길 수 있을까요?

#### 2. `DB 커넥션` 문제
- 친구 요청을 보내기 전에 사전 검증하는 과정`(사용자 존재여부 확인, 차단 유무 확인, 친구상태 확인)`이 많다보니 DB에 접근하는 부분이 너무 많아졌는데요. 사용자가 많아지면 `병목현상`이 발생하는 지점이 될 수 있을 것 같아서 어떻게 하면 줄일 수 있을까 혼자서 고민을 계속 해봤는데 마땅히 답이 나오지가 않아서요; 혹시 해결책을 위한 조언을 구할 수 있을까요?

Commit:
[#26] Alarm Create 로직 수정

- 알람을 보낼 때 to, from에 해당하는 값을 반대로 입력해서 `알람이
  반대로` 나오는 부분 수정

f-lab-dev commented:
거꾸로 생각하면 알람 저장하는데에 실패하면 친구요청 부분을 롤백을 해야할텐데 필요하지 않을까요??

f-lab-dev commented on src/main/resources/sql/alarms.sql:
```diff
@@ -0,0 +1,10 @@
+(
+    id        bigint auto_increment
+        primary key,
+    alarmTo   varchar(20)                               not null,
+    alarmFrom varchar(20)                               not null,
```
이 부분은 검색조건으로 사용될 것 같은데 인덱스를 걸어주면 어떨까요?

msugo1 replied:
인덱스를 걸어줬었는데 깜빡하고 sql 파일에 추가를 안했었네요. 업데이트 해서 push 하겠습니다.

Commit:
[#26] 피드백 반영

- 알람 저장이 실패하면 친구요청을 함께 롤백시켜주기 위해서
  `@Transactional` 어노테이션 추가

- 검색조건을 위한 인덱스 추가 in alarms 테이블

msugo1 committed:
Merge branch 'develop' into feature/26
